tests/cpp/include/test_core_op.h

```diff
@@ -311,68 +339,166 @@ class CoreOpExecutor : public test::op::OperatorDataInitializer<DType>
+        CHECK_EQ(output_types.size(), inferred_num_outputs);
+      } else {
+        if (backward_for_op) {
+          if (bwd_node_ptr) {
+            CHECK_EQ(bwd_node_ptr->inputs.size(), num_inputs);
+            input_types.resize(bwd_node_ptr->inputs.size(), -1);
+            for (size_t i = 0; i < num_inputs; ++i) {
+              const int map_key = bwd_node_ptr->inputs[i].index;
+              CHECK(index2array.find(map_key) != index2array.end());
+              const int dtype = index2array[map_key]->dtype();
+              input_types[i] = dtype;
+            }
+            for (const auto &fwd_inp : backward_for_op->inputs()) {
+              const int dtype = fwd_inp.data().type_flag_;
+              output_types.emplace_back(dtype);
+            }
+          } else {
+            for (size_t x = 0; x < num_inputs; ++x) {
+              input_types.emplace_back(default_dtype());
+            }
+            for (const auto &fwd_inp : backward_for_op->inputs()) {
+              const int dtype = fwd_inp.data().type_flag_;
+              output_types.emplace_back(dtype);
+            }
+          }
+        } else {
+          CHECK(false);  // above always true?
+          for (size_t x = 0; x < num_inputs; ++x) {
+            input_types.emplace_back(default_dtype());
+          }
+          for (size_t x = 0; x < inferred_num_outputs; ++x) {
+            output_types.emplace_back(default_dtype());
+          }
+        }
+      }
 
-      for (size_t i = 0; i < static_cast<size_t>(num_inputs); ++i) {
-        CHECK_LT(i, static_cast<int>(shapes.size()));
-        inputs_.emplace_back(i < inputs.size() ? inputs[i] : CreateRandArray(shapes[i],
-                                                                          ctx_.run_ctx.ctx));
-        inputs_p.emplace_back(&*inputs_.rbegin());
+      // Output arrays
+      if (outputs_.empty()) {
+        std::vector<nnvm::TShape> output_shapes;
+        static auto& finfer_shape = Op::GetAttr<nnvm::FInferShape>("FInferShape");
+        if (finfer_shape.count(op_)) {
+          nnvm::FInferShape call_infer_shapes = finfer_shape[op_];
+          output_shapes.resize(inferred_num_outputs);
+          call_infer_shapes(attrs_, &input_shapes, &output_shapes);
+          input_shapes_ = input_shapes;
+        } else {
+          if (backward_for_op) {
+            // BWD Input shapes
+            if (bwd_node_ptr) {
+              input_shapes.clear();
+              CHECK_EQ(bwd_node_ptr->inputs.size(), num_inputs);
+              for (size_t i = 0; i < num_inputs; ++i) {
+                const int map_key = bwd_node_ptr->inputs[i].index;
+                CHECK(index2array.find(map_key) != index2array.end());
+                const nnvm::TShape &shp = index2array[map_key]->shape();
+                input_shapes.push_back(shp);
+                const nnvm::TShape ss = input_shapes[i];
+              }
+            } else {
+              // TODO(cjolivier)
```

Comment:
Please create an issue and print a message/throw an error here

Reply:
This else block isn't needed